### PR TITLE
SAA-1208 adding a new transaction handler around attendance changes to guarantee attendances are updated before events are emitting attendance amended events.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
A problem has shown itself whereby (bulk) attendance amendment events are being processed by the sync service before the changes had fully committed in our service DB.  This resulted in the activities service and NOMIS being out of sync with each.

The crux of the changes here involve wrapping the attendance changes in a completely independent new transaction which has to complete after which we publish the events in the service/component and not use the attendance entity listener.

Hopefully this is a short term solution with regards to raising attendance amended events until we have time to revisit and address  a more complete solution towards the raising of events in general.  Ideally we should be looking to move away from entity listeners e.g. use of `TransactionalEventListener` which can determine when a transaction has committed so could replace the use of entity listeners perhaps.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AttendanceEntityListener.kt [hunk 1]
```diff
@@ -28,15 +27,6 @@ class AttendanceEntityListener {
     )
   }
 
-  @PostUpdate
-  fun onUpdate(attendance: Attendance) {
-    send(
-      OutboundEvent.PRISONER_ATTENDANCE_AMENDED,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AttendanceEntityListener.kt:34]: We now have to raise this manually when we make any changes to existing attendance records.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledInstanceService.kt [hunk 2]
```diff
@@ -62,16 +63,24 @@ class ScheduledInstanceService(
...
     log.info("Uncancelling scheduled instance $id")
 
-    val scheduledInstance = repository.findById(id)
-      .orElseThrow { EntityNotFoundException("Scheduled Instance $id not found") }
+    val uncancelledInstance = transactionHandler.new {
+      val scheduledInstance = repository.findById(id)
+        .orElseThrow { EntityNotFoundException("Scheduled Instance $id not found") }
 
-    scheduledInstance.uncancelSessionAndAttendances()
+      scheduledInstance.uncancelSessionAndAttendances()
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledInstanceService.kt:70]: This isn't quite right.  Need to return the attendances that have been uncancelled as well.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledInstanceService.kt [hunk 3]
```diff
@@ -80,21 +89,29 @@ class ScheduledInstanceService(
...
+      val scheduledInstance = repository.findById(instanceId)
+        .orElseThrow { EntityNotFoundException("Scheduled Instance $instanceId not found") }
 
-    scheduledInstance.cancelSessionAndAttendances(
-      reason = scheduleInstanceCancelRequest.reason,
-      by = scheduleInstanceCancelRequest.username,
-      cancelComment = scheduleInstanceCancelRequest.comment,
-      cancellationReason = attendanceReasonRepository.findByCode(AttendanceReasonEnum.CANCELLED),
-    )
+      scheduledInstance.cancelSessionAndAttendances(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledInstanceService.kt:96]: This isn't quite right.  Need to return the attendances that have been cancelled as well.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/TransactionHandler.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.transaction.annotation.Transactional
+
+/**
+ * To be used in services or components where a new transaction is required to ensure any DB changes are committed prior
+ * to taking any further action e.g. emitting new/update/delete events.
+ */
+@Component
+class TransactionHandler {
+  @Transactional(propagation = Propagation.REQUIRES_NEW)
+  fun <T> new(block: () -> T): Result<T> = runCatching { block() }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/TransactionHandler.kt:14]: I think the `runCatching` block will always prevent the transaction being rolled back?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/TransactionHandler.kt:14]: The thing which causes the rollback is a run time exception.

Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/TransactionHandler.kt:14]: 🤔 you might be right there will take out to be safe, easy enough to do.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesService.kt [hunk 5]
```diff
@@ -35,25 +39,33 @@ class AttendancesService(
...
+          newIssuePayment = updateRequest.issuePayment,
+          newIncentiveLevelWarningIssued = updateRequest.incentiveLevelWarningIssued,
+          newCaseNoteId = updateRequest.mayBeCaseNote(attendance.prisonerNumber)?.caseNoteId,
+          newOtherAbsenceReason = updateRequest.otherAbsenceReason,
+        )
+
+        attendanceRepository.saveAndFlush(attendance)
+      }.map { it.attendanceId }
+    }.getOrThrow()
+
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesService.kt:63]: Not sure what the `getOrThrow()` does here?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesService.kt:63]: I like the idea of this block being transactional - but its hard to follow.  If it works though, happy to live with it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/ActivitiesChangedEventHandler.kt [hunk 6]
```diff
@@ -56,9 +62,15 @@ class ActivitiesChangedEventHandler(
...
+        transactionHandler.new {
+          allocationRepository.findByPrisonCodeAndPrisonerNumber(event.prisonCode(), event.prisonerNumber())
+            .suspendPrisonersAllocations(now, event)
+            .suspendPrisonersFutureAttendances(now, event)
+        }.let { updatedAttendances ->
+          updatedAttendances.forEach {
+            outboundEventsService.send(OutboundEvent.PRISONER_ATTENDANCE_AMENDED, it.attendanceId)
+          }.also { log.info("Sending attendance amended events.") }
+        }
       }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/ActivitiesChangedEventHandler.kt:74]: Did this bit need to change? Only dealing with a single prisoner's allocations and attendances.  For tomorrow!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesService.kt (hunk 5 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesService.kt:63]: There is no getOrThrow now as taken the result out.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AttendanceIntegrationTest.kt [hunk 7]
```diff
@@ -46,9 +45,6 @@ class AttendanceIntegrationTest : IntegrationTestBase() {
   @Autowired
   private lateinit var attendanceRepository: AttendanceRepository
 
-  @Autowired
-  private lateinit var attendancesService: ManageAttendancesService
-
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AttendanceIntegrationTest.kt:51]: Unused?

**Review by timharrison-moj — COMMENTED** [review]
LGTM. Would like to test it locally (as discussed).

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AttendanceIntegrationTest.kt (hunk 7 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AttendanceIntegrationTest.kt:51]: Yes.  I suspect a hangover from something else and probably not removed.

**Review by timharrison-moj — APPROVED** [review]
LGTM. Matt demonstrated locally.
